dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderTestSuite.scala

```diff
@@ -1,118 +1,429 @@
+    }
+
+    "should be accepted if user doesn't have enough Waves to pay fee, but he take waves from order result " in {
+      val amount   = 100.waves
+      val price    = 0.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, SELL)(
+        amount -> price
+      )
+
+      node.waitOrderProcessed(wavesUsdPair, node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, BUY, amount, price, fixedFee)).message.id)
+
+      node.accountBalances(account.toAddress.toString)._1 should be(amount - fixedFee)
+    }
+
+    "should be rejected if the price is too low for completely filling by current opened orders (BUY)" in {
+      val amount      = 100.waves
+      val marketPrice = price - 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, SELL)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, BUY, amount, marketPrice, fixedFee)),
+                                 tooLowPrice("buy", "0.3"))
+    }
+
+    "should be rejected if the price is too high for completely filling by current opened orders (SELL)" in {
+      val amount      = 100.waves
+      val marketPrice = price + 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, BUY)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, SELL, amount, marketPrice, fixedFee)),
+                                 tooHighPrice("sell", "0.5"))
+    }
+
+    "should be rejected if amount of the buy market order more then user could buy" in {
+      val amount   = 101.waves
+      val price    = 1.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, SELL)(
+        amount -> price
+      )
+
+      assertBadRequestAndResponse(node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, BUY, amount, price, fixedFee)), "")
+    }
+
+    "should be rejected if user has enough balance to fill market order, but has not enough balance to pay fee in another asset" in {
+      docker.restartNode(node, ConfigFactory.parseString(s"waves.dex.order-fee.fixed.asset = $BtcId\nwaves.dex.order-fee.mode = $FIXED"))
+
+      val amount   = 10.waves
+      val price    = 1.usd
+      val transfer = 10.usd
+
+      val account = createAccountWithBalance { transfer -> Some(UsdId.toString) }
+
+      node.waitOrderStatus(wavesUsdPair,
+                           node.placeOrder(bob, wavesUsdPair, SELL, amount, price, fixedFee, feeAsset = IssuedAsset(BtcId), version = 3).message.id,
+                           "Accepted")
+
+      assertBadRequestAndResponse(
+        node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, BUY, amount, price, fixedFee, feeAsset = IssuedAsset(BtcId))),
+        "")
+    }
   }
```

Comment:
add case
orderbook:
asks:
1@1
1@2
1@3
market order 3@2

Reply:
https://github.com/wavesplatform/dex/blob/dex-88-tests/dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderTestSuite.scala#L375